Firestore/core/src/firebase/firestore/remote/stream.mm

```diff
@@ -0,0 +1,318 @@
+                                           : absl::string_view{};
+  }();
+
+  grpc_stream_ = CreateGrpcStream(datastore_, token);
+  grpc_stream_->Start();
+}
+
+void Stream::OnStreamStart() {
+  EnsureOnQueue();
+
+  state_ = State::Open;
+  DoOnStreamStart();
+}
+
+// Backoff
+
+void Stream::BackoffAndTryRestarting() {
+  EnsureOnQueue();
+
+  LOG_DEBUG("%s backoff", GetDebugDescription());
+
+  HARD_ASSERT(state_ == State::Error,
+              "Should only perform backoff in an error case");
+
+  state_ = State::Backoff;
+  backoff_.BackoffAndRun([this] { ResumeStartFromBackoff(); });
+}
+
+void Stream::ResumeStartFromBackoff() {
+  EnsureOnQueue();
+
+  HARD_ASSERT(state_ == State::Backoff, "Backoff elapsed but state is now: %s",
+              state_);
+
+  state_ = State::Initial;
+  Start();
+  HARD_ASSERT(IsStarted(), "Stream should have started.");
+}
+
+void Stream::InhibitBackoff() {
+  EnsureOnQueue();
+
+  HARD_ASSERT(!IsStarted(),
+              "Can only cancel backoff in a stopped state (was %s)", state_);
+
+  // Clear the error condition.
+  state_ = State::Initial;
+  ResetBackoff();
+}
+
+void Stream::ResetBackoff() {
+  EnsureOnQueue();
+  backoff_.Reset();
+}
+
+// Idleness
+
+void Stream::MarkIdle() {
+  EnsureOnQueue();
+
+  if (IsOpen() && !idleness_timer_) {
+    idleness_timer_ = worker_queue_->EnqueueAfterDelay(
+        kIdleTimeout, idle_timer_id_, [this] { Stop(); });
+  }
+}
+
+void Stream::CancelIdleCheck() {
+  EnsureOnQueue();
+  idleness_timer_.Cancel();
+}
+
+// Read/write
+
+void Stream::OnStreamRead(const grpc::ByteBuffer& message) {
+  EnsureOnQueue();
+
+  HARD_ASSERT(IsStarted(), "OnStreamRead called for a stopped stream.");
+
+  if (bridge::IsLoggingEnabled()) {
+    LOG_DEBUG("%s headers (whitelisted): %s", GetDebugDescription(),
+              Datastore::GetWhitelistedHeadersAsString(
+                  grpc_stream_->GetResponseHeaders()));
+  }
+
+  Status read_status = DoOnStreamRead(message);
+  if (!read_status.ok()) {
+    grpc_stream_->Finish();
+    // Don't expect GRPC to produce status -- since the error happened on the
+    // client, we have all the information we need.
+    OnStreamError(read_status);
+    return;
+  }
+}
+
+// Stopping
+
+void Stream::Stop() {
+  EnsureOnQueue();
+
+  LOG_DEBUG("%s stop", GetDebugDescription());
+
+  if (!IsStarted()) {
+    return;
+  }
+
+  RaiseGeneration();
+
+  // If the stream is in the auth stage, GRPC stream might not have been created
+  // yet.
+  if (grpc_stream_) {
+    LOG_DEBUG("%s Finishing GRPC stream", GetDebugDescription());
+    FinishGrpcStream(grpc_stream_.get());
+    ResetGrpcStream();
+  }
+
+  // If this is an intentional close ensure we don't delay our next connection
+  // attempt.
+  ResetBackoff();
+
+  // State must be updated before calling the delegate.
+  state_ = State::Initial;
+  // Stopping the stream was initiated by the client, so we have all the
+  // information we need.
+  DoOnStreamFinish(Status::OK());
```

Comment:
The events coming out of streams involve Open and Close. This should be named NotifyStreamClose to make it clear that we're specifically giving the stream an extension point for notifying the delegate.